duckduckgo/iOS
# NetP test & interruption notifications

Task/Issue URL: https://app.asana.com/0/0/1205601804599806/f
BSK PR: https://github.com/duckduckgo/BrowserServicesKit/pull/516

**Description**:

This PR adds a Connection Interruption failure simulation option and updates BSK to bring in the changes made to allow this. It also adds the relevant `UNUserNotificationCenter` handling for sending notifications from netP, with a basic test notification to test this (do we need this after adding the interruption test?).

I also added some changes to the PacketTunnelProvider in order to test connection interruptions for the iOS NetP Notifications project. I found that simply setting reasserting to true causes the ConnectionTester to trigger an interruption, then setting it to false recovers it. This has the benefit of testing the result of the real callback, not needing to mess with any of the connection tester’s callback handling, not needing to add any new state and also mimicking an actual interruption (as the VPN connection status also changes).

**Steps to test this PR**:
1. Build this branch through Xcode
2. Start Network Protection (Settings -> Network Protection)
3. **Observe the notifications permission prompt**. Agree to this.
4. Go to the Debug menu and simulate a connection interruption (Settings -> Debug -> Network Protection -> Simulate Connection Interruption)
5. **Observe the interruption notification**
6. Go back to Network Protection settings
7. **Observe the status is reconnecting**
8. Wait for and **observer the reconnected notification.**
9. Go back to the Debug menu and send a test notification
10. **Observe the test notification**

**Designs:**

https://www.figma.com/file/zuCBX2UpJpS8CgtcmpU1hf/Network-Protection-for-iOS?type=design&node-id=252-21944&mode=design&t=QlNmmkcLYiIemlBB-0

Added screenshots to the[ PR task](https://app.asana.com/0/0/1205601804599806/f) too.

**Copy Testing**:


**Orientation Testing**:


**Device Testing**:


**OS Testing**:


**Theme Testing**:


—
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
